src/off-chain/historical-data/cowEndpointSurplus.py

```diff
@@ -0,0 +1,258 @@
+"""
+EBBO Historical Data Testing via block number inputs or a single settlement hash.
+Uses CoW Endpoint provided callData.
+"""
+import json
+import requests
+from fractions import Fraction
+import directory
+
+
+class EBBOHistoricalDataTesting:
+    def __init__(self):
+        self.total_orders = 0
+        self.higher_surplus_orders = 0
+        self.total_surplus_eth = 0
+        self.file_name = str
+
+        self.solver_dict = {
+            "1Inch": [0, 0],
+            "Raven": [0, 0],
+            "0x": [0, 0],
+            "PLM": [0, 0],
+            "Quasilabs": [0, 0],
+            "Otex": [0, 0],
+            "SeaSolver": [0, 0],
+            "Laertes": [0, 0],
+            "ParaSwap": [0, 0],
+            "BalancerSOR": [0, 0],
+            "BaselineSolver": [0, 0],
+            "Legacy": [0, 0],
+            "NaiveSolver": [0, 0],
+            "DMA": [0, 0],
+            "CowDexAg": [0, 0],
+            "Barter": [0, 0],
+        }
+
+    def total_surplus(
+        self, start_block=None, end_block=None, settlement_hash=None, file_name=None
+    ):
+        self.file_name = file_name
+
+        if settlement_hash is not None:
+            self.get_order_surplus(settlement_hash)
+
+        elif start_block is not None and end_block is not None:
+            # Etherscan endpoint call for settlements between start and end block
+            try:
+                etherscan_url = f"https://api.etherscan.io/api?module=account&action=txlist&address=0x9008D19f58AAbD9eD0D60971565AA8510560ab41&startblock={start_block}&endblock={end_block}&sort=desc&apikey={directory.ETHERSCAN_KEY}"
+                # all "result" go into results (based on API return value names from docs)
+                settlements = json.loads((requests.get(etherscan_url)).text)["result"]
+
+                # loop over all settlements and get surplus difference for each
+                for settlement in settlements:
+                    self.get_order_surplus(settlement["hash"])
+                self.statistics_output(start_block, end_block)
+            except ValueError:
+                print("etherscan error.")
```

Comment:
I would organise the logic a bit differently:

1. Have an empty list of settlement hashes.
2. if settlement_hash is not None -> add a single hash to the list above.
3. elif start_block is not None and end_block is not None -> fetch all hashes and add them to the list above.
4. Once we have all the hashes, have a separate function that receives a list of hashes and returns a list of solver_competition data by calling cow api.
5. For each solver_competition data, call `get_order_surplus`